### ISSUE TITLE
Is there a way to change the ip in the blacklist dynamically after the program run？like read a config from a file，and the file is changing always。 

### ISSUE THREAD

**Issue by ltaini** [basil00/Divert#154]

**Comment by TechnikEmpire** [comment]
https://www.learn-c.org/

**Comment by ltaini** [comment]
> https://www.learn-c.org/

Sorry,i meant do i need to call WinDivertOpen every time the filter is changed。

**Comment by TechnikEmpire** [comment]
I'm using strong language here with "doing it wrong" and "against the design". You can do things this way but frankly the idea is to enable you to write code handling the packets as if you were writing driver code but without the huge burden of figuring out how to do that properly or sign drivers.

If you knew how the underlying packet filtering engine in the OS worked you'd understand that it's supported but "against the design" to keep pushing and popping filter matching parameters like Yosemite Sam pops off rounds.

**Comment by ltaini** [comment]
> Filters are per-handle, and the handle is your pipe into the driver working for you to give you packets that match your initial filter.
> 
> You cannot modify it after and shouldn't. You should create a static number of handles that are generic enough to match all conditions you want to handle, and then have the dynamic part your after be done by logic in your own code.
> 
> That's a long way of saying yes you would need to keep opening a new handle every time you want to change the filter, but also that this would be doing it wrong so don't do it.
> 
> You're pushing and pulling things to and from kernel space. Aside from your proposed use case being against the overall design, you're putting a burden on the driver to maintain resources. Resources are more difficult to come by and manage in kernel land.
> 
> Tldr: yes but don't do it

The current requirement is that  my application will provide sevice as http server to receive configs ,and will be called anytime ,and i need to change the filter real-time。
Do you mean that in that case i need to prepare a number of filter handles？but in this case the number should be unlimited？
So，do you have any suggestion in this case？
By the way，i have a question that if i want to block a hostname like www.XXX.com .Is block udp packet and get host name from requests the right way to meet this requirement?
Truely thanks.

**Comment by TechnikEmpire** [comment]
No it's not. Look at my projects CitadelCore.Windows and HttpFilteringEngine. My sole purpose for using WinDivert is to block porn so if that's your goal then pilfer through my projects because there's a lot of well developed solutions for that problem.

It sounds like you want to run a remote filtering proxy. CitadelCore.Windows is a local filtering proxy. If you want to create a service you need proxy configurations, not WinDivert filter strings per customer.

I think your confusion comes from not differentiating between packet filter layer and http filter layer. You only need a single filter string to grab the appropriate packets for filtering http/s. They are not the same later.

You're looking at the undercarriage of the car trying to figure out what down there will make the interior look different. They're related but worlds apart.

Use WinDivert to capture all TCP traffic and pipe it to a filtering proxy server and then let that server have authorization and loads different configurations for your customers and thus loads different http filtering rules. On that note, my project DistillNET is a simple but stupidly fast url filtering/matching engine based on adblock plus syntax. Like I said poke around in my projects, but your issue is not a WinDivert issue.

**Comment by ltaini** [comment]
> In fact if you want a jump start selling filtering software, just fork these guys:
> 
> https://github.com/cloudveiltech/FilterManager
> 
> https://github.com/cloudveiltech/Filter-Windows
> 
> I wrote all that software. They've butchered it pretty bad because they don't know what they're doing but yeah, that's a complete content filter for windows with a server side manager where you can create and edit customer/user accounts and upload filtering rules/lists etc etc. Just watch out, those people like to steal software and violate copyright licenses so don't publish anything anywhere they can see it.
> 
> It, like mostly all my work is MPL2 so you can fork it and keep your new file additions closed-source for profit.

Okay,thank you so much for your help .I will take a look as you said.It's so kind of you^.^
In fact , i am just doing something just for passing test instead of selling it ))

**Comment by basil00** [comment]
Dynamically changing the filter is hard to design.  The problem is that the internal packet queue needs to be "refiltered" allowing queued packets match the old but not the new filter to be "re-injected".  This queue can be quite long, so is not a quick operation we can hold a spinlock for.  Furthermore, the refiltering must work in parallel with new packet processing, as well as any additional "change filter" request that may occur at any time.  Another problem is that the required callouts might need to be changed.  So it is a mess, and ultimately the overheads may just be the same as opening a new handle anyway.

**Comment by ltaini** [comment]
> Dynamically changing the filter is hard to design. The problem is that the internal packet queue needs to be "refiltered" allowing queued packets match the old but not the new filter to be "re-injected". This queue can be quite long, so is not a quick operation we can hold a spinlock for. Furthermore, the refiltering must work in parallel with new packet processing, as well as any additional "change filter" request that may occur at any time. Another problem is that the required callouts might need to be changed. So it is a mess, and ultimately the overheads may just be the same as opening a new handle anyway.

Got it^.^ Thanks for your explanation.
